addons/mail/controllers/discuss/translation.py

```diff
@@ -0,0 +1,57 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import requests
+from odoo.http import request, route, Controller
+
+
+# TODO KFR: Refacto (abstraction) + Errors in response
+class DiscussTranslationController(Controller):
+    languages = dict()
+
+    @route("/discuss/translation/translate", type="json", auth="user")
+    def translate(self, body):
+        result = dict.fromkeys(["source", "value", "errors"])
+        if source_code := self._detect_source(body):
+            result.update({
+                "value": self._get_translation(body, source_code),
+                "source": self._format_source(source_code)  
+            })
+        
+        return result
+            
+    def _prepare_request(self, endpoint):
+        api_key = request.env['ir.config_parameter'].sudo().get_param('discuss.message_translation_api_key')
+        url = f"https://translation.googleapis.com/language/translate/v2{endpoint}/?key={api_key}"
+        return {"url": url, "headers": {}}
+            
+    def _post_request(self, endpoint="", payload={}):
+        params = self._prepare_request(endpoint)
+        params["headers"]["Content-Type"] = "application/json"
+        if response := requests.post(params["url"], json=payload, headers=params["headers"]):
+            return response.json()
+
+    def _get_request(self, endpoint="", payload={}):
+        params = self._prepare_request(endpoint)
+        if response := requests.get(params["url"], params=payload, headers=params["headers"]):
+            return response.json()
+        
+    def _fetch_languages(self, source_code):
+        if result := self._get_request("/languages", {"target": request.env.user.lang[:2]}):
+            for data in result.get("data", {}).get("languages", []):
+                self.languages.update({data["language"]: data["name"]})
+
+    def _format_source(self, source_code):
+        if not self.languages:
+            self._fetch_languages(source_code)
+        if self.languages:
+                return self.languages.get(source_code)
+
+    def _detect_source(self, body):
+        if source := self._post_request("/detect", {"q": body}):
+            return source.get("data", {}).get("detections", [[{}]])[0][0].get("language")
+        
+    def _get_translation(self, body, source_code):
+        payload = {"q": body, "target": request.env.user.lang[:2], "source": source_code}
+
+        if result := self._post_request(payload=payload):
+            return result.get("data", {}).get("translations", [{}])[0].get("translatedText") 
```

Comment:
Haven't looked at the code in depth, but at a glance there's some heavy function split caused by subtle duplication of `_detect_source` and `_get_translation`, both of which are called only once and in sequence.

If you group them together, I think the code could be much simpler (less functions to call), and it can actually be more performant because you don't have to call some code twice.

Reply:
@aka-odoo Could you have a look at the simplified version and give me some feedback ? I wanted to make a separate commit to ease the review but I squashed the new refactor commit by mistake 🤦‍♂️

Reply:
@alexkuhn

Reply:
This is much better! Thanks! 😊